### PR TITLE
Authorization levels

### LINKED ISSUE #26: Implement authorization levels
Implement authorization levels to add admins or moderators.

### PR DESCRIPTION
**Pull request by ZuluNovember**
Closes #26

## Add authorization levels

- Add role field to `User` model.
- Create `isAdmin` , `isModerator` middlewares.

**Intended to be used in admin/moderator endpoints.**

### CODE REVIEW

File: src/middlewares/accessControl.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+const isAdmin = async (req, res, next) => {
+  try {
+    const user = User.findById(req.user.id);
+    if (user.role === 'admin') {
+      next();
+    } else {
+      res.status(400).json({
+        status: 'Fail',
+        message: 'You are trying to access an admin route',
```
Comment (Kishimoto96) [inline src/middlewares/accessControl.js:11]: For the messages, I think just showing a 404 error is enough without specifying what layer is in need, this is what I see generally on websites when I try to access unauthorized pages.

Screen Shot 2022-02-19 at 11 27 56 PM

File: src/middlewares/accessControl.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+const isModerator = async (req, res, next) => {
+  try {
+    const user = User.findById(req.user.id);
+    if (user.role === 'admin' || user.role === 'moderator') {
+      next();
+    } else {
+      res.status(400).json({
+        status: 'Fail',
+        message: 'You are trying to access a moderator route',
```
Comment (Kishimoto96) [inline src/middlewares/accessControl.js:27]: here also :)

**Review by Kishimoto96 — APPROVED** [review]
Super clean work as always!

File: src/middlewares/accessControl.js (hunk 1 above)
Comment (ZuluNovember) [reply src/middlewares/accessControl.js:11]: Hey, @Kishimoto96. Could you clarify what _**without specifying what layer is in need**_ means? Also this 404 error is what you receive by trying to access a front-end page. It's not a request to API. What do you think about [403](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/403)?

Comment (Kishimoto96) [reply src/middlewares/accessControl.js:11]: Hi @ZuluNovember ,
I mean we don't have to let the user know which kind of authentication is needed to access any private page, and yeah just giving a 403 Forbidden error would be perfect, Thanks :)

**Review by ibraheemhajjar — APPROVED** [review]
good job @ZuluNovember i think its all good

File: src/middlewares/accessControl.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    if (user.role === 'admin') {
+      next();
+    } else {
+      res.status(403).json({
+        status: 'Fail',
+        message: 'You are trying to access an admin route',
+      });
+    }
+  } catch (err) {
+    res.status(500).json(err);
```
Comment (Shrreya) [inline src/middlewares/accessControl.js:15]: Also add a logger call to log this error.

File: src/middlewares/accessControl.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    if (user.role === 'admin' || user.role === 'moderator') {
+      next();
+    } else {
+      res.status(403).json({
+        status: 'Fail',
+        message: 'You are trying to access a moderator route',
+      });
+    }
+  } catch (err) {
+    res.status(500).json(err);
```
Comment (Shrreya) [inline src/middlewares/accessControl.js:31]: A logger call here too.

**Review by Shrreya — CHANGES_REQUESTED** [review]

**Pull request merged**